### DISCUSSION TITLE
Improve the readability of function definition

### DISCUSSION THREAD

**Discussion by thangnbn** [microsoft/pylance-release#3186] [Enhancement]
This is inspired from the behavior of the Java extension. So currently, for example, if my mouse hovers on a function, a docstring popup as the following displays:

![python-docstring](https://i.imgur.com/qsKu0Re.png)

I think it is unslightly, compared to "docstring" I get from the Java extension below, and not to mention there are many docstring styles, which creates inconsistency.

![java-docstring](https://i.imgur.com/lGFG0LR.png)

This may relate to microsoft/vscode-python#6676. I think we can simply parse the docstrings before displaying them - implementing different parsers for different styles. I'm happy to work on this issue.

**Comment by bschnurr** [comment]
current implementation just tries to reproduce the plaintext style in markdown.

docstirng conversion is in pyright.  https://github.com/microsoft/pyright/blob/ed13daf16f2621600a0f1ca67fdd18aa1b79d98b/packages/pyright-internal/src/analyzer/docStringConversion.ts#L558

**Reply by thangnbn** [reply]
Is it possible to implement this as another extension? - when users install this extension, the docstring popup will be overwritten by the new extension? So far I don't know there is any method in vscode API to overwrite other extensions' hover provider.

**Reply by rchl** [reply]
Is there a particular reason why you've linked to the source code @bschnurr? Is that to encourage someone to improve it? Or are you not interested in improving that and you wouldn't accept a PR?

I also think that current un-formatted presentation is not very readable and could use a bit of processing to make parameter names stand out.

**Reply by erictraut** [reply]
If you want to attempt to improve this code, I recommend first writing a proposal in the form of a small design spec in this thread. We can then review and discuss the proposal and make sure we're on the same page before code is written/modified.

**Reply by thangnbn** [reply]
@erictraut Is there any template/guideline for the short proposal? And where should I put the proposal for revision?

**Reply by erictraut** [reply]
I was thinking that you could just describe the proposal in this thread.

**Reply by thangnbn** [reply]
So ideally, the extension will provide a common function hover no matter what docstring style the user chooses. I was looking at this link for some docstring examples [](https://queirozf.com/entries/python-docstrings-reference-examples) and the behavior of PyCharm. For example, if the given function is as followed:

```
def func(arg1, arg2):
    """Summary line.

    Extended description of function.

    :param int arg1: Description of arg1.
    :param str arg2: Description of arg2.
    :raise: ValueError if arg1 is equal to arg2
    :return: Description of return value
    :rtype: bool

    :example:

    >>> a=1
    >>> b=2
    >>> func(a,b)
    True
    """

    if arg1 == arg2:
        raise ValueError('arg1 must not be equal to arg2')

    return True
```

In PyCharm, the function hover is as followed:

<img width="324" alt="Screenshot 2022-12-23 at 8 06 11 PM" src="https://user-images.githubusercontent.com/54808964/209344016-456de611-9a3b-46f7-bdb8-c23aaecde021.png">

I think the analyzer will take the docstring and convert it into a common form. In this example, I propose our function hover will be (in markdown) as the following template:

```
Summary line.

Extended description of function

### Params
- arg1: Description of arg1.
- arg2: Description of arg2.

### Return
Description of return value

### Raises
- ValueError if arg1 is equal to arg2

### Examples
>>> a=1
>>> b=2
>>> func(a,b)
True
```

We will omit any section (params, return, raises, examples) if it is not available, and perhaps we will ignore other information - you can see that in the above example we skip the `:rtype:` annotation. Looking at the referred code, I think we can further break the _parseFieldList state into more states like _parseParams, _parseReturn, etc. for processing.

**Reply by thangnbn** [reply]
Also, the Basic Inline Markup section in the [epytext documentation](https://epydoc.sourceforge.net/manual-epytext.html) mentions some specific markup that the style assumes. Perhaps we want to render that as well in our analyzer implementation.

**Reply by rchl** [reply]
This would require quite a bit more work. I suppose an alternative could be to just prettify lines a bit without grouping. So for example turn

```
:param foo: Some param
:param bar: Other param
:returns: Something
```

to:

_Param_ **foo** Some param
_Param_ **bar** Other param
_Returns_ Something

Or some variation of that. That would be much simpler to do and would still prove more readable than plain sphinx syntax.

**Reply by thangnbn** [reply]
I'm thinking without grouping, how can we process the Google's docstring style, which the sample docstring is as the following because the current implementation simply looks at each line in the string.

```py
def function_with_types_in_docstring(param1, param2):
    """Example function with types documented in the docstring.

    `PEP 484`_ type annotations are supported. If attribute, parameter, and
    return types are annotated according to `PEP 484`_, they do not need to be
    included in the docstring:

    Args:
        param1 (int): The first parameter.
        param2 (str): The second parameter.

    Returns:
        bool: The return value. True for success, False otherwise.

    .. _PEP 484:
        https://www.python.org/dev/peps/pep-0484/

    """
    pass
```

**Reply by rchl** [reply]
Would this need to be processed? It's already very readable as plain text.

**Reply by thangnbn** [reply]
To be honest, I like to have a common template like what Pycharm does - Pycharm converts all docstring styles into the same popup I screenshot above. However, if nobody really cares about that, I agree that we better off just do the work for Sphinx (the :param: style) and Epytext (the @param style).
